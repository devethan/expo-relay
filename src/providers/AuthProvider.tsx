import React, { useReducer } from 'react';
import AsyncStorage from '@react-native-community/async-storage';
import { Environment } from 'relay-runtime';
import Relay from '../relay';
import { User } from '../types';
import createCtx from '../utils/createCtx';

interface Context {
  state: State;
  setUser(user: User): void;
  resetUser(): void;
}
const [useCtx, Provider] = createCtx<Context>();

export enum ActionType {
  ResetUser = 'reset-user',
  SetUser = 'set-user',
}

export interface State {
  user: User | null;
  relay: Environment;
}

const initialState: State = {
  user: null,
  relay: Relay.environment,
};

interface SetUserAction {
  type: ActionType.SetUser;
  payload: User;
}

interface ResetUserAction {
  type: ActionType.ResetUser;
}

type Action = SetUserAction | ResetUserAction;

interface Props {
  children?: React.ReactElement;
}

type Reducer = (state: State, action: Action) => State;

const setUser = (dispatch: React.Dispatch<SetUserAction>) => (
  user: User,
): void => {
  dispatch({
    type: ActionType.SetUser,
    payload: user,
  });
};

const resetUser = (dispatch: React.Dispatch<ResetUserAction>) => (): void => {
  Relay.init();

  AsyncStorage.removeItem('@UserStorage:login_token').then(() => {
    dispatch({
      type: ActionType.ResetUser,
    });
  });
};

const reducer: Reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'reset-user':
      return { ...initialState, relay: Relay.environment };
    case 'set-user':
      return { ...state, user: action.payload };
    default:
      return state;
  }
};

function AuthProvider(props: Props): React.ReactElement {
  const [state, dispatch] = useReducer<Reducer>(reducer, initialState);

  const actions = {
    setUser: setUser(dispatch),
    resetUser: resetUser(dispatch),
  };

  return <Provider value={{ state, ...actions }}>{props.children}</Provider>;
}

const AuthContext = {
  useAuthContext: useCtx,
  AuthProvider,
};

export { useCtx as useAuthContext, AuthProvider };
export default AuthContext;
