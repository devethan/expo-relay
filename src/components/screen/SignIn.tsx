import type {
  SignInEmailMutation,
  SignInEmailMutationResponse,
} from '__generated__/SignInEmailMutation.graphql';
import { graphql, useMutation } from 'react-relay/hooks';
import styled, { css } from 'styled-components/native';
import AsyncStorage from '@react-native-community/async-storage';
import { AuthStackNavigationProps } from '../navigation/AuthStackNavigator';
import Button from '../shared/Button';
import Constants from 'expo-constants';
import React from 'react';
import { useAuthContext } from '../../providers/AuthProvider';

const Container = styled.View`
  flex: 1;
  background-color: ${({ theme }): string => theme.background};
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const WebCss = css`
  cursor: pointer;
`;
const MobileCss = css`
  /* cursor: pointer; */
`;

const StyledImage = styled.Image`
  border-radius: 15px;
  align-self: center;
  margin: 25px;
  ${Constants.platform?.web ? WebCss : MobileCss};
`;

const StyledTextInput = styled.TextInput`
  width: 320px;
  height: 40px;
  align-self: center;
  border-width: 1.3px;
  border-color: lightgray;
  border-radius: 3px;
  padding: 10px;
  margin-bottom: 5px;
  color: ${({ theme }): string => theme.fontColor};
`;

const StyledText = styled.Text`
  width: 320px;
  text-align: center;
  margin: 10px 0;
`;

const SignupText = styled(StyledText)`
  color: #373d78;
  text-decoration: underline;
`;

const ErrorMessage = styled(StyledText)`
  color: #f57b51;
`;

interface Props {
  navigation: AuthStackNavigationProps<'SignIn'>;
}

const SignInEmail = graphql`
  mutation SignInEmailMutation($email: String!, $password: String!) {
    signInEmail(email: $email, password: $password) {
      token
      user {
        id
        email
        name
        photoURL
      }
    }
  }
`;

function SignIn(props: Props): React.ReactElement {
  const { setUser } = useAuthContext();
  const [email, setEmail] = React.useState<string>('ethan1@test.com');
  const [password, setPassword] = React.useState<string>('test');
  const [error, setError] = React.useState<string>('');

  const [commit, isInFlight] = useMutation<SignInEmailMutation>(SignInEmail);

  const mutationConfig = {
    variables: {
      email,
      password,
    },
    onCompleted: async (response: SignInEmailMutationResponse): void => {
      const { token, user } = response.signInEmail;
      await AsyncStorage.setItem('@UserStorage:login_token', token);

      setUser({
        ...user,
      });
    },
    onError: (error): void => {
      console.error(error);
      setError('Check your email and password');
    },
  };

  const handleClick = (): void => {
    setError('');
    commit(mutationConfig);
  };

  const moveScreen = (): void => {
    props.navigation.navigate('SignUp');
  };

  React.useEffect(() => {
    console.log('[AuthStack] mounted');
  }, []);

  return (
    <Container>
      <StyledImage
        style={{ width: 150, height: 150 }}
        source={{
          uri: 'https://avatars2.githubusercontent.com/u/45788556?s=200&v=4',
        }}
      />
      <StyledTextInput
        testID="input-email"
        value={email}
        onChangeText={(value: string): void => setEmail(value)}
        textContentType="emailAddress"
        placeholder="email"
      />
      <StyledTextInput
        testID="input-password"
        value={password}
        onChangeText={(value: string): void => setPassword(value)}
        textContentType="password"
        secureTextEntry={true}
        placeholder="Password"
      />
      <ErrorMessage testID="error-text" numberOfLines={1}>{error}</ErrorMessage>
      <Button
        testID="btn-signin"
        onClick={handleClick}
        text={'SignIn'}
        isLoading={isInFlight}
      />
      <SignupText onPress={moveScreen}>
        {"Don't have an account yet?"}
      </SignupText>
    </Container>
  );
}

export default SignIn;
