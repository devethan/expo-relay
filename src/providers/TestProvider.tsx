import React, { ReactElement, Suspense } from 'react';
import { RelayMockEnvironment, createMockEnvironment } from 'relay-test-utils';
import { ThemeProvider, ThemeType } from '../providers/ThemeProvider';

import { AuthProvider } from './AuthProvider';
import { Environment } from 'relay-runtime';
import ErrorBoundary from '../ErrorBoundary';
import { RelayEnvironmentProvider } from 'react-relay/hooks';
import SuspenseScreen from '../components/screen/Suspense';

interface Props {
  initialThemeType?: ThemeType;
  children?: React.ReactElement;
}

interface RelayProvidersProps {
  children?: React.ReactElement;
}

export const environment: RelayMockEnvironment = createMockEnvironment();

const RelayProviderWrapper = ({
  children,
}: RelayProvidersProps): ReactElement => {
  return (
    <RelayEnvironmentProvider environment={environment}>
      <ErrorBoundary fallback={<SuspenseScreen error />}>
        <Suspense fallback={<SuspenseScreen />}>{children}</Suspense>
      </ErrorBoundary>
    </RelayEnvironmentProvider>
  );
};

// Add providers here
const RootProvider = ({
  initialThemeType = ThemeType.LIGHT,
  children,
}: Props): React.ReactElement => {
  return (
    <ThemeProvider initialThemeType={initialThemeType}>
      <AuthProvider>
        <RelayProviderWrapper>{children}</RelayProviderWrapper>
      </AuthProvider>
    </ThemeProvider>
  );
};

export default RootProvider;
