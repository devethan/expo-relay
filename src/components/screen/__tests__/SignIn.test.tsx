import 'react-native';

import React, { ReactElement } from 'react';
import {
  RenderResult,
  fireEvent,
  render,
  wait,
  waitForElement,
} from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import AsyncStorage from '@react-native-community/async-storage';
import AuthContext from '../../../providers/AuthProvider';
import { MockPayloadGenerator } from 'relay-test-utils';
import SignIn from '../SignIn';
import { ThemeType } from '../../../providers/ThemeProvider';
import { act } from 'react-test-renderer';
import { environment } from '../../../providers/TestProvider';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: ReactElement;
let testingLib: RenderResult;

describe('[SignIn] rendering test', () => {
  beforeEach(() => {
    props = createTestProps();
    component = createTestElement(<SignIn {...props} />);
  });

  it('should render without crashing', async () => {
    component = createTestElement(<SignIn {...props} />);
    testingLib = render(component);

    // Remove snapshot testing for now for issue https://github.com/VirgilSecurity/virgil-e3kit-js/issues/82
    expect(testingLib.baseElement).toMatchSnapshot();
  });

  it('should render [Dark] mode without crashing', () => {
    component = createTestElement(<SignIn {...props} />, ThemeType.DARK);
    testingLib = render(component);
    expect(testingLib.baseElement).toBeTruthy();
  });
});

describe('[SignIn] interaction', () => {
  beforeAll(() => {
    props = createTestProps();
    component = createTestElement(<SignIn {...props} />);
    testingLib = render(component);
  });

  it('should invoke changeText event handler when email changed', async () => {
    const emailInput = testingLib.getByTestId('input-email');
    await waitForElement(() => emailInput); // why we should use it?

    act(() => {
      fireEvent.changeText(emailInput, 'email@email.com');
    });

    expect(emailInput.props.value).toEqual('email@email.com');
  });

  it('should invoke changeText event handler when password changed', async () => {
    const passwordInput = testingLib.getByTestId('input-password');
    await waitForElement(() => passwordInput); // why we should use it?

    act(() => {
      fireEvent.changeText(passwordInput, 'password');
    });

    expect(passwordInput.props.value).toEqual('password');
  });

  describe('about signin submit action', () => {
    beforeAll(() => {
      testingLib = render(component);
      jest.spyOn(AsyncStorage, 'setItem').mockImplementation(jest.fn());
      jest.spyOn(AuthContext, 'useAuthContext').mockImplementation(() => ({
        state: {
          user: undefined,
          relay: environment,
        },
        setUser: jest.fn().mockReturnValue({
          id: 'userId',
          email: 'email@email.com',
          nickname: 'nickname',
          statusMessage: 'status',
        }),
        resetUser: jest.fn(),
      }));
    });
    it('should make request to signin when button has pressed and navigation switchs to [MainStack]', async () => {
      const emailText = testingLib.getByTestId('input-email');
      await waitForElement(() => emailText);

      act(() => {
        fireEvent.changeText(emailText, 'email@email.com');
      });

      const passwordText = testingLib.getByTestId('input-password');
      await waitForElement(() => passwordText);

      act(() => {
        fireEvent.changeText(passwordText, 'password');
      });

      const signinBtn = testingLib.getByTestId('btn-signin');
      await waitForElement(() => signinBtn);

      act(() => {
        fireEvent.press(signinBtn);
        environment.mock.resolveMostRecentOperation((operation) =>
          MockPayloadGenerator.generate(operation),
        );
      });
    });

    it('should fail request to signin when button has pressed with wrong info', async () => {
      const emailText = testingLib.getByTestId('input-email');
      await waitForElement(() => emailText);

      act(() => {
        fireEvent.changeText(emailText, 'wrong-email@email.com');
      });

      const passwordText = testingLib.getByTestId('input-password');
      await waitForElement(() => passwordText);

      act(() => {
        fireEvent.changeText(passwordText, 'wrong-password');
      });

      const signinBtn = testingLib.getByTestId('btn-signin');
      await waitForElement(() => signinBtn);

      act(() => {
        fireEvent.press(signinBtn);
        environment.mock.rejectMostRecentOperation(new Error('reject'));
      });

      const errorText = testingLib.getByTestId('error-text');
      await act(() => wait());
      expect(errorText).toBeTruthy();
    });
  });

  it('should navigate to [SignUp] when button has pressed', async () => {
    const signUpBtn = testingLib.getByText("Don't have an account yet?");
    await waitForElement(() => signUpBtn);

    act(() => {
      fireEvent.press(signUpBtn);
    });

    expect(props.navigation.navigate).toHaveBeenCalledWith('SignUp');
  });
});
