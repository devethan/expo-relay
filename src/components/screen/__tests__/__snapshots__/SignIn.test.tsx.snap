// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SignIn] rendering test should render without crashing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://avatars2.githubusercontent.com/u/45788556?s=200&v=4",
          }
        }
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderRadius": 15,
              "marginBottom": 25,
              "marginLeft": 25,
              "marginRight": 25,
              "marginTop": 25,
            },
            Object {
              "height": 150,
              "width": 150,
            },
          ]
        }
      />
      <TextInput
        allowFontScaling={true}
        placeholder="email"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderColor": "lightgray",
              "borderRadius": 3,
              "borderWidth": 1.3,
              "color": "black",
              "height": 40,
              "marginBottom": 5,
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
              "width": 320,
            },
          ]
        }
        testID="input-email"
        textContentType="emailAddress"
        underlineColorAndroid="transparent"
        value="ethan1@test.com"
      />
      <TextInput
        allowFontScaling={true}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderColor": "lightgray",
              "borderRadius": 3,
              "borderWidth": 1.3,
              "color": "black",
              "height": 40,
              "marginBottom": 5,
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
              "width": 320,
            },
          ]
        }
        testID="input-password"
        textContentType="password"
        underlineColorAndroid="transparent"
        value="test"
      />
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#f57b51",
              "marginBottom": 10,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 10,
              "textAlign": "center",
              "width": 320,
            },
          ]
        }
        testID="error-text"
      >
        
      </Text>
      <TouchableOpacity
        activeOpacity={0.5}
        testID="btn-signin"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#069ccd",
                "borderColor": "#069ccd",
                "borderRadius": 4,
                "borderWidth": 2,
                "justifyContent": "center",
                "marginBottom": 2,
                "marginLeft": 2,
                "marginRight": 2,
                "marginTop": 2,
                "minHeight": 35,
                "width": 320,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 14,
                },
              ]
            }
          >
            SignIn
          </Text>
        </View>
      </TouchableOpacity>
      <Text
        style={
          Array [
            Object {
              "color": "#373d78",
              "marginBottom": 10,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 10,
              "textAlign": "center",
              "textDecorationColor": "black",
              "textDecorationLine": "underline",
              "textDecorationStyle": "solid",
              "width": 320,
            },
          ]
        }
      >
        Don't have an account yet?
      </Text>
    </View>
  </View>
</View>
`;
